### PR TITLE
Add alpine supervisor build info

### PR DESCRIPTION
**Pull request by nghiant2710**
This change is [Reviewable](https://reviewable.io/reviews/resin-io/resin-supervisor/134)

### CODE REVIEW

**Comment by Page-** [comment]
Is there a reason we aren't directly switching the default to alpine? I'd rather not have to deal with testing both debian based and alpine based supervisors if we don't have to

**Comment by nghiant2710** [comment]
@Page- Alpine based supervisors need more testing to make sure they work properly.
- [ ] Need to test on amd64 devices (nuc) since I don't have that board.
- [ ] Need to test all supervisor endpoints (using SDK), I only tested with functions on dashboard.
I will spend more time to test this but It would be great if someone from our team who has a nuc can test amd64 alpine supervisor.
When alpine based supervisors are fully tested, we can retire debian based ones.

One more thing, we still have to keep armel debian supervisor since alpine doesn't support that architecture atm.

**Comment by nghiant2710** [comment]
@Page- @pcarranzav jenkins build results here:
https://jenkins.dev.resin.io/job/armhf-alpine-supervisor-gh/21/
https://jenkins.dev.resin.io/job/i386-alpine-supervisor-gh/4/
https://jenkins.dev.resin.io/job/amd64-alpine-supervisor-gh/10/

I've not added these jobs to supervisor master job since it might delay other build. Will add after this PR merged

**Comment by nghiant2710** [comment]
@Page- @pcarranzav jenkins build results here:
https://jenkins.dev.resin.io/job/armhf-alpine-supervisor-gh/21/
https://jenkins.dev.resin.io/job/i386-alpine-supervisor-gh/4/
https://jenkins.dev.resin.io/job/amd64-alpine-supervisor-gh/10/

I've not added these jobs to supervisor master job since it might delay other build. Will add after this PR merged

File: package.json [hunk 1]
```diff
@@ -25,7 +25,7 @@
     "resin-register-device": "^2.0.0",
     "request-progress": "^0.3.1",
     "rwlock": "^5.0.0",
-    "sqlite3": "~3.0.4",
+    "sqlite3": "3.0.9",
```
Comment (Page-) [inline package.json]: You took away the `~`/`^`?

File: Dockerfile.amd64-alpine [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+COPY config/supervisor/ /etc/supervisor/
+
+# Install dependencies
+WORKDIR /app
+COPY package.json postinstall.sh /app/
+RUN set -x \
+	&& buildDeps='g++ sqlite-dev sqlite-libs make' \
+	&& apk add --update $buildDeps \
+	&& JOBS=MAX npm install --unsafe-perm --production --no-optional \
+	&& cd /app/node_modules/sqlite3 && ./node_modules/.bin/node-pre-gyp install --build-from-source --sqlite=/usr/lib \
```
Comment (Page-) [inline Dockerfile.amd64-alpine]: Why is this needed?

File: entry.alpine.sh [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+fi
+
+export DBUS_SYSTEM_BUS_ADDRESS="unix:path=/mnt/root/run/dbus/system_bus_socket"
+
+/usr/bin/supervisord -c /etc/supervisor/supervisord.conf
+
+supervisorctl -c /etc/supervisor/supervisord.conf start resin-supervisor
+supervisorctl -c /etc/supervisor/supervisord.conf start go-supervisor
+
+tail -f /var/log/supervisor/supervisord.log
```
Comment (Page-) [inline entry.alpine.sh]: We don't have this in our standard images do we?  I'm wondering as I think this will cause docker logs to just grow and grow by default, as it doesn't do log rotation or anything that I'm aware of

File: entry.alpine.sh [hunk 4]
```diff
@@ -0,0 +1,33 @@
+#!/bin/sh
+
+set -e
+
+[ -d /dev/net ] ||
+    mkdir -p /dev/net
+[ -c /dev/net/tun ] ||
+    mknod /dev/net/tun c 10 200
+
+cd /app
```
Comment (Page-) [inline entry.alpine.sh]: Aren't we already in `/app` due to the `WORKDIR` command?

File: Dockerfile.amd64-alpine (hunk 2 above)
Comment (pcarranzav) [inline Dockerfile.amd64-alpine]: There's an incompatibility between the libsqlite from debian and alpine, so the prebuilt binaries will not work here.

Comment (Page-) [inline Dockerfile.amd64-alpine]: Is it possible to pass `--build-from-source` to `npm install`? If we can it'd be nice to do that to avoid any other incompatibilities that might happen as well

File: package.json (hunk 1 above)
Comment (nghiant2710) [inline package.json]: Yup, only v3.0.9 contains the fix we need for amd64 alpine supervisor.

File: entry.alpine.sh (hunk 3 above)
Comment (nghiant2710) [inline entry.alpine.sh]: This is from the standard images (debian supervisor). @pcarranzav can we improve this?

File: entry.alpine.sh (hunk 4 above)
Comment (nghiant2710) [inline entry.alpine.sh]: yeah, this can be removed.

File: package.json (hunk 1 above)
Comment (Page-) [inline package.json]: So `3.0.10`/`3.1.0` remove the fix?  Because both `~3.0.9` or `^3.0.9` will force `3.0.9` as the minimum version

File: entry.alpine.sh (hunk 3 above)
Comment (Page-) [inline entry.alpine.sh]: Oh right, I thought this was our supervisor's log file, it's the supervisord log file in reality, this is ok (and needed to stop the container instantly exiting)

File: package.json (hunk 1 above)
Comment (pcarranzav) [inline package.json]: Yeah apparently there was a patch that was then rolled back.

Comment (Page-) [inline package.json]: Well that's annoying, ok, a fixed version works then

File: Dockerfile.amd64-alpine (hunk 2 above)
Comment (pcarranzav) [inline Dockerfile.amd64-alpine]: Oh, apparently it's possible: https://github.com/mapbox/node-pre-gyp/issues/58#issuecomment-39261084
@nghiant2710 wanna give it a try?

File: entry.alpine.sh (hunk 4 above)
Comment (nghiant2710) [inline entry.alpine.sh]: Removed!

File: Dockerfile.amd64-alpine (hunk 2 above)
Comment (nghiant2710) [inline Dockerfile.amd64-alpine]: @pcarranzav, I've tried and had no issue at all. PR updated!
Jenkins builds passed!
https://jenkins.dev.resin.io/job/amd64-alpine-supervisor-gh/13/
https://jenkins.dev.resin.io/job/armhf-alpine-supervisor-gh/25/
https://jenkins.dev.resin.io/job/i386-alpine-supervisor-gh/7/

File: Makefile [hunk 5]
```diff
@@ -1,6 +1,6 @@
 DISABLE_CACHE = 'false'
 
-ARCH = rpi# rpi/amd64/i386/armv7hf/armel
+ARCH = rpi# rpi/amd64/i386/armv7hf/armel/armhf-alpine/amd64-alpine/i386-alpine
```
Comment (pcarranzav) [inline Makefile]: @nghiant2710 can we use the same architecture here and have the alpine be passed as a separate parameter?
e.g. we use `ARCH=rpi` and `BASE_DISTRO=-alpine`, and then:
- Use `${ARCH}${BASE_DISTRO}` to choose the Dockerfile
- keep using `ARCH` for the image name
- append `BASE_DISTRO` to the tag

So that we have images that look like `resin/amd64-supervisor:1.7.0-alpine`.

This would correspond to what we had discussed privately with @Page- and @dfunckt

Comment (pcarranzav) [inline Makefile]: We'd also have to use separate builds for rpi and armv7hf even though it's the same thing, but I don't see that as too much of a problem.

File: Dockerfile.armhf-alpine [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+	&& npm dedupe \
+	&& npm cache clean \
+	&& rm -rf /tmp/* \
+	&& apk del $buildDeps \
+	&& rm -rf /var/cache/apk/*
+
+# Copy source
+COPY . /app/
+
+RUN mv -f /app/entry.alpine.sh /app/entry.sh
```
Comment (Page-) [inline Dockerfile.armhf-alpine]: Can we instead change the `ENTRYPOINT` and `ln -sf /app/entry.sh /start` entries below to point to the alpine entry, rather than replacing the other

File: Makefile (hunk 5 above)
Comment (nghiant2710) [inline Makefile]: Sure, no problem at all

File: Dockerfile.armhf-alpine (hunk 6 above)
Comment (nghiant2710) [inline Dockerfile.armhf-alpine]: yeah, makes sense.
I've updated the PR

File: Makefile (hunk 5 above)
Comment (nghiant2710) [inline Makefile]: @pcarranzav, @Page- I've update PR. Distro moved to the tag.

**Comment by pcarranzav** [comment]
:lgtm: but let's wait for @Page- 's approval too, and give this last version a test @nghiant2710

File: automation/jenkins_build.sh [hunk 7]
```diff
@@ -11,36 +11,82 @@ docker pull resin/${ARCH}-supervisor:${ESCAPED_BRANCH_NAME} || docker pull resin
...
-	JOB_NAME=${JOB_NAME} \
-	DEPLOY_REGISTRY=registry.resinstaging.io/ \
-	PUBNUB_SUBSCRIBE_KEY=${PUBNUB_SUBSCRIBE_KEY} \
-	PUBNUB_PUBLISH_KEY=${PUBNUB_PUBLISH_KEY} \
-	MIXPANEL_TOKEN=${MIXPANEL_TOKEN} \
-	deploy
+if [ -z "$BASE_DISTRO" ]; then
+	# Debian
+	# Build the images
+	make SUPERVISOR_VERSION=${ESCAPED_BRANCH_NAME} \
```
Comment (Page-) [inline automation/jenkins_build.sh]: Can you do:
```bash
MAKE_ARGS="ARCH=${ARCH} \
	JOB_NAME=${JOB_NAME} \
	BASE_DISTRO= \
	PUBNUB_SUBSCRIBE_KEY=${PUBNUB_SUBSCRIBE_KEY} \
	PUBNUB_PUBLISH_KEY=${PUBNUB_PUBLISH_KEY} \
	MIXPANEL_TOKEN=${MIXPANEL_TOKEN}"
```
and then just
```bash
make ${MAKE_ARGS} \
	SUPERVISOR_VERSION=${ESCAPED_BRANCH_NAME} \
	DEPLOY_REGISTRY= \
	deploy
```
to make it easier to see the difference between the different commands (and reduce duplication).

File: automation/jenkins_build.sh [hunk 8]
```diff
@@ -11,36 +11,82 @@ docker pull resin/${ARCH}-supervisor:${ESCAPED_BRANCH_NAME} || docker pull resin
...
-	deploy
-make SUPERVISOR_VERSION=${VERSION} \
-	ARCH=${ARCH} \
-	JOB_NAME=${JOB_NAME} \
-	DEPLOY_REGISTRY=registry.resinstaging.io/ \
-	PUBNUB_SUBSCRIBE_KEY=${PUBNUB_SUBSCRIBE_KEY} \
-	PUBNUB_PUBLISH_KEY=${PUBNUB_PUBLISH_KEY} \
-	MIXPANEL_TOKEN=${MIXPANEL_TOKEN} \
-	deploy
+if [ -z "$BASE_DISTRO" ]; then
```
Comment (Page-) [inline automation/jenkins_build.sh]: It should be fine to just do: `make ... BASE_DISTRO=${BASE_DISTRO} ...` I think, after all if `$BASE_DISTRO` is empty that's equivalent to doing `make ... BASE_DISTRO= ...` anyway

File: Makefile [hunk 9]
```diff
@@ -9,13 +10,20 @@ JOB_NAME = 1
...
-SUPERVISOR_IMAGE=$(DEPLOY_REGISTRY)$(IMAGE)
-
 PUBNUB_SUBSCRIBE_KEY = sub-c-bananas
 PUBNUB_PUBLISH_KEY = pub-c-bananas
 MIXPANEL_TOKEN = bananasbananas
 
+ifeq ($(BASE_DISTRO),alpine)
+	IMAGE = "resin/$(ARCH)-supervisor:$(SUPERVISOR_VERSION)-$(BASE_DISTRO)"
+	DOCKERFILE = alpine.$(ARCH)
+else
```
Comment (Page-) [inline Makefile]: I think it would be better to change this to a not exists check, as it stands if I specify `BASE_DISTRO=fedora` for instance then it will just silently use a debian base without any warning/error

File: automation/jenkins_build.sh (hunk 8 above)
Comment (nghiant2710) [inline automation/jenkins_build.sh]: Hmm, it didn't remember why I need to check the base_distro here but you're right. Will remove now

File: automation/jenkins_build.sh (hunk 7 above)
Comment (nghiant2710) [inline automation/jenkins_build.sh]: Updated!

File: Makefile (hunk 9 above)
Comment (nghiant2710) [inline Makefile]: @Page- this is exactly what you want. We need to check if it's `alpine` or not to put `alpine` in the tag. Otherwise, it uses debian.

Comment (Page-) [inline Makefile]: My problem is that you can say `make BASE_DISTRO=fedora` or `make BASE_DISTRO=alpien` and it will just silently use debian with no warning or anything, which would be very confusing when you're expecting it to be `fedora` or `alpine` (but typoed)

Comment (nghiant2710) [inline Makefile]: Added a check with info message, if `BASE_DISTRO` exists then build alpine supervisor, otherwise debian supervisor will be built.

**Comment by Page-** [comment]
SGTM
